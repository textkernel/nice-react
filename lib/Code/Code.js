import React, { Component } from 'react';
import PropTypes from 'prop-types';
import niceProps from '../props';

export default class Code extends Component {
    static propTypes = {
        context: PropTypes.oneOf(niceProps.codeContexts),
        size: PropTypes.oneOf(niceProps.codeSizes),
        heading: PropTypes.string,
        children: PropTypes.oneOfType([
            PropTypes.node,
            PropTypes.string
        ]),
        icon: PropTypes.string,
        syntax: PropTypes.string
    }

    static defaultProps = {
        context: null,
        size: niceProps.codeSizes[0],
        heading: null,
        children: null,
        icon: null,
        syntax: null
    }

    renderCodeHeading() {
        const { context, heading, icon, syntax } = this.props;

        if (!heading && !syntax) {
            return null;
        }

        const classes = ['nice-code-heading'];

        if (context && niceProps.codeContexts.indexOf(context) > -1) {
            classes.push(context);
        }
        if (icon) {
            classes.push(icon);
        }

        return (<div className={ classes.filter(c => !!c).join(' ') }>
            { heading }
            { 
                syntax 
                ? <div className="nice-code-syntax">
                    { syntax }
                </div>
                : null
            }
        </div>);
    }

    render() {
        const { children, size } = this.props;

        const classes = [];
        if (size) {
            classes.push(size);
        }

        return (<div className="nice-code-container">
            { this.renderCodeHeading() }
            <code className={ classes.join(' ') }>
                { children }
            </code>
        </div>);
    }
}
