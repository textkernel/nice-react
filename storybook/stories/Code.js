import React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, text, select } from '@storybook/addon-knobs';
import { withInfo } from '@storybook/addon-info';
import props from '../../src/props';

import { Code } from 'nice-react';

const stories = storiesOf('Code Blocks', module);

stories.addDecorator((story, context) => withInfo('')(story)(context));
stories.addDecorator(withKnobs);

stories.add('Default', () => {
    const { codeSizes: sizes, codeContexts: contexts } = props;
    const heading = text('Heading', 'Sample API response');
    const syntax = text('Syntax', 'JSON');
    const size = select('Height', sizes, sizes[0]);
    const context = select('Context', contexts, contexts[0]);

    return (<Code heading={ heading } icon="icon fa-code" syntax={ syntax } context={ context } size={ size }>
        { `{
    "$schema": "http://json-schema.org/draft-04/schema#",
    "title": "Product set",
    "type": "array",
    "items": {
        "title": "Product",
        "type": "object",
        "properties": {
            "id": {
                "description": "The unique identifier for a product",
                "type": "number"
            },
            "name": {
                "type": "string"
            },
            "price": {
                "type": "number",
                "minimum": 0,
                "exclusiveMinimum": true
            },
            "tags": {
                "type": "array",
                "items": {
                    "type": "string"
                },
                "minItems": 1,
                "uniqueItems": true
            },
            "dimensions": {
                "type": "object",
                "properties": {
                    "length": {"type": "number"},
                    "width": {"type": "number"},
                    "height": {"type": "number"}
                },
                "required": ["length", "width", "height"]
            },
            "warehouseLocation": {
                "description": "Coordinates of the warehouse with the product",
                "$ref": "http://json-schema.org/geo"
            }
        },
        "required": ["id", "name", "price"]
    }
}` }
    </Code>);
});
