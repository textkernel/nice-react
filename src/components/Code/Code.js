import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import niceProps from '../../props';
import './Code.scss';

class Code extends Component {
    renderCodeHeading() {
        const { context, heading, icon, syntax } = this.props;

        if (!heading && !syntax) {
            return null;
        }

        const classes = ['nice-code-heading'];

        if (context && niceProps.codeContexts.indexOf(context) > -1) {
            classes.push(context);
        }
        if (icon) {
            classes.push(icon);
        }

        return (<div className={ classes.filter(c => !!c).join(' ') }>
            { heading }
            { 
                syntax 
                ? <div className="nice-code-syntax">
                    { syntax }
                </div>
                : null
            }
        </div>);
    }

    render() {
        const { children, size, className, style } = this.props;

        const classes = [];
        if (size) {
            classes.push(size);
        }

        return (
            <div className={classnames('nice-code-container', className)} style={style} >
                { this.renderCodeHeading() }
                <code className={ classes.join(' ') }>
                    { children }
                </code>
            </div>
        );
    }
}

Code.propTypes = {
    className: PropTypes.string,
    context: PropTypes.oneOf(niceProps.codeContexts),
    size: PropTypes.oneOf(niceProps.codeSizes),
    heading: PropTypes.string,
    children: PropTypes.oneOfType([
        PropTypes.node,
        PropTypes.string
    ]),
    icon: PropTypes.string,
    syntax: PropTypes.string,
    style: PropTypes.object,
};

Code.defaultProps = {
    className: null,
    context: null,
    size: niceProps.codeSizes[0],
    heading: null,
    children: null,
    icon: null,
    syntax: null,
    style: {},
};

export default Code;
