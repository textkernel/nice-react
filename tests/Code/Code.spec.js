import React from 'react';
import Code from '../../lib/Code';

describe('<Code />', () => {
    it('should render a Code block without heading', () => {
        const wrapper = shallow(<Code>
            { `{
    "some": "json"
}` }
        </Code>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render a Code block with heading but no syntax', () => {
        const wrapper = shallow(<Code heading="Sample code block">
            { `{
    "some": "json"
}` }
        </Code>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render a Code block with heading and syntax', () => {
        const wrapper = shallow(<Code heading="Sample code block" syntax="JSON">
            { `{
    "some": "json"
}` }
        </Code>);
        expect(wrapper).toMatchSnapshot();
    });

    it('should render a Code block with custom context, size and icon', () => {
        const wrapper = shallow(<Code context="good" heading="Custom context" icon="icon fa-code" size="restrain-m">
            { `{
    "some": "json"
}` }
        </Code>);
        expect(wrapper).toMatchSnapshot();
    });
});
