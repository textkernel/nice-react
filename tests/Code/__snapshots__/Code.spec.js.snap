// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Code /> should render a Code block with custom context, size and icon 1`] = `
<div
  className="nice-code-container"
  style={Object {}}
>
  <div
    className="nice-code-heading good icon fa-code"
  >
    Custom context
  </div>
  <code
    className="restrain-m"
  >
    {
    "some": "json"
}
  </code>
</div>
`;

exports[`<Code /> should render a Code block with heading and syntax 1`] = `
<div
  className="nice-code-container"
  style={Object {}}
>
  <div
    className="nice-code-heading"
  >
    Sample code block
    <div
      className="nice-code-syntax"
    >
      JSON
    </div>
  </div>
  <code
    className="restrain-s"
  >
    {
    "some": "json"
}
  </code>
</div>
`;

exports[`<Code /> should render a Code block with heading but no syntax 1`] = `
<div
  className="nice-code-container"
  style={Object {}}
>
  <div
    className="nice-code-heading"
  >
    Sample code block
  </div>
  <code
    className="restrain-s"
  >
    {
    "some": "json"
}
  </code>
</div>
`;

exports[`<Code /> should render a Code block without heading 1`] = `
<div
  className="nice-code-container"
  style={Object {}}
>
  <code
    className="restrain-s"
  >
    {
    "some": "json"
}
  </code>
</div>
`;

exports[`<Code /> should render with a Code block with custom className/style 1`] = `
<div
  className="nice-code-container my-custom-classname"
  style={
    Object {
      "left": 50,
    }
  }
>
  <div
    className="nice-code-heading good icon fa-code"
  >
    Hello
  </div>
  <code
    className="restrain-s"
  >
    { some": "json" }
  </code>
</div>
`;
